File: dgs-users/src/main/java/com/github/arhor/dgs/users/api/listener/UserRelationalEventListener.java
package com.github.arhor.dgs.users.api.listener;

import com.github.arhor.dgs.users.config.props.AppProps;
import com.github.arhor.dgs.users.data.entity.UserEntity;
import io.awspring.cloud.sns.core.SnsNotification;
import io.awspring.cloud.sns.core.SnsOperations;
import jakarta.annotation.Nonnull;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.relational.core.mapping.event.AbstractRelationalEventListener;
import org.springframework.data.relational.core.mapping.event.AfterDeleteEvent;
import org.springframework.data.relational.core.mapping.event.AfterSaveEvent;
import org.springframework.messaging.MessagingException;
import org.springframework.retry.annotation.Retryable;
import org.springframework.stereotype.Component;

import java.util.Map;

@Component
@Retryable(retryFor = MessagingException.class)
public class UserRelationalEventListener extends AbstractRelationalEventListener<UserEntity> {

    private static final String HEADER_PAYLOAD_TYPE = "xPayloadType";

    private final SnsOperations snsOperations;
    private final String userUpdatedEventsTopic;
    private final String userDeletedEventsTopic;

    @Autowired
    public UserRelationalEventListener(final SnsOperations snsOperations, final AppProps appProps) {
        this.snsOperations = snsOperations;
        this.userUpdatedEventsTopic = appProps.getAws().getSns().getUserUpdatedEvents();
        this.userDeletedEventsTopic = appProps.getAws().getSns().getUserDeletedEvents();
    }

    @Override
    @SuppressWarnings("DataFlowIssue")
    public void onAfterSave(@Nonnull final AfterSaveEvent<UserEntity> event) {
        sendNotification(
            userUpdatedEventsTopic,
            new UserStateChange.Updated(event.getEntity().getId())
        );
    }

    @Override
    public void onAfterDelete(@Nonnull final AfterDeleteEvent<UserEntity> event) {
        sendNotification(
            userDeletedEventsTopic,
            new UserStateChange.Deleted((Long) event.getId().getValue())
        );
    }

    private void sendNotification(final String destination, final UserStateChange payload) {
        snsOperations.sendNotification(
            destination,
            new SnsNotification<>(
                payload,
                Map.of(HEADER_PAYLOAD_TYPE, payload.type())
            )
        );
    }

    sealed interface UserStateChange {

        String type();

        record Updated(Long id) implements UserStateChange {
            @Override
            public String type() {
                return "UserStateChange.Updated";
            }
        }

        record Deleted(Long id) implements UserStateChange {
            @Override
            public String type() {
                return "UserStateChange.Deleted";
            }
        }
    }
}

File: dgs-users/src/main/kotlin/com/github/arhor/dgs/users/api/listener/UserRelationalEventListener.kt
package com.github.arhor.dgs.users.api.listener

import com.github.arhor.dgs.lib.HEADER_PAYLOAD_TYPE
import com.github.arhor.dgs.users.config.props.AppProps
import com.github.arhor.dgs.users.data.entity.UserEntity
import io.awspring.cloud.sns.core.SnsNotification
import io.awspring.cloud.sns.core.SnsOperations
import org.springframework.data.relational.core.mapping.event.AbstractRelationalEventListener
import org.springframework.data.relational.core.mapping.event.AfterDeleteEvent
import org.springframework.data.relational.core.mapping.event.AfterSaveEvent
import org.springframework.messaging.MessagingException
import org.springframework.retry.annotation.Retryable
import org.springframework.stereotype.Component

@Component
@Retryable(retryFor = [MessagingException::class])
class UserRelationalEventListener(
    private val snsOperations: SnsOperations,
    appProps: AppProps,
) : AbstractRelationalEventListener<UserEntity>() {

    private val userChangesTopic = appProps.aws.sns.userChanges

    override fun onAfterSave(event: AfterSaveEvent<UserEntity>) {
        sendNotification(
            payload = UserChange.Updated(id = event.entity.id!!),
            type = USER_CHANGE_UPDATED
        )
    }

    override fun onAfterDelete(event: AfterDeleteEvent<UserEntity>) {
        sendNotification(
            payload = UserChange.Deleted(id = event.id.value as Long),
            type = USER_CHANGE_DELETED
        )
    }

    private fun sendNotification(payload: UserChange, type: String) {
        snsOperations.sendNotification(
            userChangesTopic,
            SnsNotification(
                payload,
                mapOf(HEADER_PAYLOAD_TYPE to type)
            )
        )
    }

    companion object {
        private const val USER_CHANGE_UPDATED = "UserChange.Updated"
        private const val USER_CHANGE_DELETED = "UserChange.Deleted"
    }

    sealed interface UserChange {
        data class Updated(val id: Long) : UserChange
        data class Deleted(val id: Long) : UserChange
    }
}
